Mathlib/NumberTheory/ModularForms/NormTrace.lean

```diff
@@ -0,0 +1,271 @@
+
+open Pointwise in
+lemma Subgroup.conjAct_pointwise_smul_eq_self
+    {G : Type} [Group G] {H : Subgroup G} {g : G} (hg : g ∈ normalizer H) :
+    ConjAct.toConjAct g • H = H :=
+  Subgroup.conjAct_pointwise_smul_iff.2 hg
+
+local notation "𝒬" => ℋ ⧸ (𝒢.subgroupOf ℋ)
+
+instance : MulAction ℋ ℋ := Monoid.toMulAction ..
+instance : MulAction ℋ 𝒬 := .quotient ..
+
+section SlashInvariantForm
+
+variable [SlashInvariantFormClass F 𝒢 k]
+
+/-- For `f` invariant under `𝒢`, this is a function on `(ℋ ⧸ 𝒢 ⊓ ℋ) × ℍ → ℂ` which packages up the
+translates of `f` by `ℋ`. -/
+def quotientFunc (q : 𝒬) (τ : ℍ) : ℂ :=
+  q.liftOn (fun g ↦ ((f : ℍ → ℂ) ∣[k] g.val⁻¹) τ) (fun h h' hhh' ↦ by
+    obtain ⟨j, hj, hj'⟩ : ∃ g ∈ 𝒢, h' = h * g := by
+      rw [← Quotient.eq_iff_equiv, Quotient.eq, QuotientGroup.leftRel_apply] at hhh'
+      exact ⟨h⁻¹ * h', hhh', mod_cast (mul_inv_cancel_left h h').symm⟩
+    simp [hj', SlashAction.slash_mul, SlashInvariantFormClass.slash_action_eq f j⁻¹ (inv_mem hj)])
+
+lemma quotientFunc_mk (h : ℋ) : quotientFunc f ⟦h⟧ = (f : ℍ → ℂ) ∣[k] h.val⁻¹ :=
+  rfl
+
+lemma quotientFunc_smul {h} (hh : h ∈ ℋ) (q : 𝒬) :
+    quotientFunc f q ∣[k] h = quotientFunc f ((⟨h, hh⟩ : ℋ)⁻¹ • q) := by
+  induction q using Quotient.inductionOn with | h r =>
+  simp [quotientFunc_mk, SlashAction.slash_mul]
+
+lemma SlashAction.sum_slash {β G α ι : Type*} [Monoid G] [AddCommGroup α] [SlashAction β G α]
+    (k : β) (g : G) {a : ι → α} {s : Finset ι} :
+    (∑ i ∈ s, a i) ∣[k] g = ∑ i ∈ s, a i ∣[k] g := by
+  classical
+  induction s using Finset.induction_on with
+  | empty => simp
+  | insert i t hi IH => simp [hi, IH]
+
+lemma ModularForm.prod_slash {ι : Type*} (k : ℤ) (g : GL (Fin 2) ℝ)
+    {f : ι → ℍ → ℂ} {s : Finset ι} (hs : s.Nonempty) :
+    (∏ i ∈ s, f i) ∣[k * s.card] g = |g.det.val| ^ (s.card - 1) • (∏ i ∈ s, f i ∣[k] g) := by
+  classical
+  induction s using Finset.induction_on with
+  | empty => simp_all
+  | insert i t hi IH =>
+    by_cases ht : t.Nonempty
+    · have : 0 < t.card := by aesop
+      simp only [Finset.prod_insert hi, Finset.card_insert_of_notMem hi, Nat.cast_succ,
+        mul_add, mul_one, add_comm]
+      simp [IH ht, mul_slash, show t.card + 1 - 1 = t.card - 1 + 1 by omega, pow_succ,
+        ← mul_smul, mul_comm]
+    · obtain rfl : t = ∅ := by simpa using ht
+      simp
+
+lemma ModularForm.prod_fintype_slash {ι : Type*} (k : ℤ) (g : GL (Fin 2) ℝ)
+    {f : ι → ℍ → ℂ} [Fintype ι] [Nonempty ι] :
+    (∏ i, f i) ∣[k * Fintype.card ι] g =
+      |g.det.val| ^ (Fintype.card ι - 1) • (∏ i, f i ∣[k] g) := by
+  simpa using ModularForm.prod_slash k g Finset.univ_nonempty
+
+variable (ℋ)
+
+/-- The trace of a slash-invariant form, as a slash-invariant form. -/
+def SlashInvariantForm.trace [𝒢.IsFiniteRelIndex ℋ] : SlashInvariantForm ℋ k where
+  toFun := letI := Fintype.ofFinite 𝒬; ∑ q : 𝒬, quotientFunc f q
+  slash_action_eq' h hh := by
+    letI := Fintype.ofFinite 𝒬
```

Comment:
```suggestion
    let := Fintype.ofFinite 𝒬
```
("I" is for "inline", which is rarely useful in a proof; it used to be "instance" in Lean 3...)
Using `letI`/`haveI` can make sense in definitions, to obtain better terms, though.